ALBT-81 WIP: bigquery fork of sqlpad

murphyke opened the pull request:
The initial commit represents a manual rebase on sqlpad-master of hongkongkiwi's bigquery fork of sqlpad.

The usability of this is unknown.  The runQuery function clearly needs work, as do the tests.

bruth commented on server/drivers/bigquery/test.sh:
```diff
@@ -0,0 +1,23 @@
+
+if [ -z "${BIGQUERY_TEST_DATASET_LOCATION}" ]; then
+  echo "Must set BIGQUERY_TEST_DATASET_LOCATION with Google Cloud Project Region e.g. asia-southeast1"
+  exit 4
```
Any reason for the different exit codes? Is this what the author did?

murphyke replied:
Yes, that's what the author did. I noticed that but forgot to change it.

murphyke replied:
I changed the error codes to be less surprising. It's not a bad policy to have a unique error code for every unique thing that can go wrong, but it was unnecessary in this case.

bruth commented on server/drivers/bigquery/test.js:
```diff
@@ -0,0 +1,144 @@
+    });
+  });
+
+  it('implements runQuery and runs queries under a limit', function() {
```
Maybe make a note how this is "under a limit"?

murphyke replied:
OK. I want to find where in the UI or elsewhere this `maxRows` comes into play.

murphyke replied:
This is referring to SQLPad's `maxRows` parameter than can be changed in its configuration. I will push those changes tomorrow morning.

bruth commented on server/drivers/bigquery/index.js:
```diff
@@ -0,0 +1,144 @@
+    .then(([job]) => {
+      // Waits for the query to finish
+      return job.getQueryResults({
+        maxResults: connection.maxRows + 1
```
What is the `+ 1` for?

murphyke replied:
It's so the code can tell if the query would return more results than `maxRows`. I haven't explored that, but I guess it's a SQLPad feature to have that limit and be able to tell the user if the results were truncated or not.

murphyke commented:
See https://jira.arcus.chop.edu:8443/browse/ALBT-81

Byron may be in the best position to test this.

I'm not asking anyone except Byron to test, but here's the recipe:

- Create a Google Cloud project via [the Jenkins job for making a GCP project](https://jenkins-ops-dbhi.research.chop.edu/view/DevOps/job/GCP/job/terraform-gcp-project-sandbox/build?delay=0sec).
- You will get an email telling you the project ID/name, etc.
- Log into cloud.google.com, go to the console.
- You have to add the BigQuery API to your project, but I forget the details.
- Go into BigQuery (under Big Data in the list of services on the left) and create a new dataset.
- Download the service account key file from the key path given in the email.
- Clone this repo and switch to the albt-81-sqlpad-bq branch.
- Run `scripts/build.sh` (this is per DEVELOPER-GUIDE.md).
- `export SQLPAD_ADMIN=admin`
- `export SQLPAD_ADMIN_PASSWORD=something`
- `cd server && node server.js --dbPath ../db --port 3010`
- Surf to http://localhost:3010
- Create a new Connection, choosing the BigQuery driver, adding your GCP project ID, the path to the service account keyfile, the dataset you created, and the location (`US`).

bruth commented:
All the steps worked. When testing the connection, I get the response:

```
{
    "error": "Error: Invalid value at 'max_results.value' (TYPE_UINT32), \"\""
}
```

bruth commented:
Hm, but I can create the connection and run queries (including inserts) just fine! So there must be something weird with the "test connection" code.

murphyke commented:
> Hm, but I can create the connection and run queries (including inserts) just fine! So there must be something weird with the "test connection" code.

Will check.

murphyke commented:
> > Hm, but I can create the connection and run queries (including inserts) just fine! So there must be something weird with the "test connection" code.
> 
> Will check.

I fixed that. The problem was that `runQuery` did not handle the case where `maxRows` was not passed in the `connection` definition.

murphyke commented:
@bruth Aside from the `testConnection` issue, did you find anything else wrong? I seem to remember you saying something about a query not returning the right results, but I skimmed it and now can't find where you said that.

Apart from that, I think I've addressed your comments so far.

I would like to add a README.md to the bigquery driver directory describing how to test it.

murphyke commented:
This now has a README and smarter (skippable) tests.  Also, unqualified table names in statements are now interpreted as being in the connection's dataset. That last one had been nagging at me; the nodejs bigquery api did not document any "default dataset" behavior, but I knew it had to be there. After enough reading and head-banging, I figured it out.

bruth commented:
> I seem to remember you saying something about a query not returning the right results

I tried this again and this seems like a strange one.. I can reproduce if the query takes like >7 seconds. Executing the same query again will hit the BQ cache and then the results render. It's possible this is a general SQLPad issue and not BQ specific. Is there any kind of timeout that would prevent rendering after some period of time? That would be bizarre, but the only thing I can think of at the moment.

To reproduce connect to that BQ streams dataset and do some kind of aggregation (like a count) that takes a decent amount of time.

murphyke commented:
> To reproduce connect to that BQ streams dataset and do some kind of aggregation (like a count) that takes a decent amount of time.

Looking at this.

murphyke commented:
> > To reproduce connect to that BQ streams dataset and do some kind of aggregation (like a count) that takes a decent amount of time.
> 
> Looking at this.

@bruth I found the problem. If `maxRows` (called `maxResults` in the API) is specified at all (even with a huge value), the first query always comes back empty with no errors, but repeating the same exact query always succeeds, as you pointed out. I've double-checked what the nodejs client API is doing by snooping on the REST API traffic (requests and responses).

This _seems_ like a BigQuery API bug, but that's hard to believe, and I will double-check. There is no matching issue in the Google Issue Tracker (search terms: "bigquery  maxresults").

TODO:

1. Get a second opinion from `bq` and/or the BigQuery console.
2. Push changes so the driver will simply ignore the `maxRows` parameter, possibly making our fork usable.
3. Maybe try to reproduce with a public dataset and report the problem, if it seems real.

bruth commented:
Interesting finding. Anything I can help with or test?

dmead commented:
gonna review this but i'm reading up on the big query api first.

murphyke commented:
> Interesting finding. Anything I can help with or test?

@bruth OK, I think I _actually_ figured it out this time.

From testing with `bq`, I can see that it is polling for the results (it polls 13 times).

The sample query code in the `@google-cloud/bigquery` package (for Node.js) does not do this; the comment on the relevant statement just says "wait for the results".

The docs for that function describe a `timeoutMs` option and say that the default timeout is 0. However, the latter is not true; it _does_ wait, I think for 10 seconds.

But if you specify the `timeoutMs` option, you can make the timeout as large as you want.

So it seems our driver could either poll or just make the one function call with a large timeout specified.

The SQLPad backend has a tuneable timeout with a default of 300 seconds.

It would also be possible to expose a user-configurable timeout in the connection settings and use that, but if it were larger than the SQLPad backend's timeout, it would be ignored.

@bruth What are your thoughts on the best arrangement?

The first option that occurs to me is to make the SQLPad backend timeout value "infinitely large" and let the user control the timeout in the connection settings, with a default of 5 minutes, maybe.

Or we could just lock the two timeouts at the same value and see how it plays with users.

In the future, another possibility is that after the timeout, we could throw a custom error message (which gets displayed prominently in the query results area), giving the user a URL at which to collect their results, but in addition to writing a little service, we would have to grapple with the associated workflow for the user.

bruth commented:
5 minutes should be more than enough for queries. Since the backend will be BigQuery by default and for anything large, I and Bill have not observed query times exceeding 2 minutes at this point.

So if the easiest thing is to increase the `timeoutMs` to match the backend timeout, then let's do that. Less configuration is better if we can help it.

bruth commented:
I tested this and confirmed that a 17 second query returned on the first try 🙌

murphyke commented:
I am going to push a couple of additional changes.

- Fix to tests: my `before` hook (to set up tables) was asynchronous, but I wasn't using Mocha's `done` callback.
- If the timeout is reached, instead of just returning nothing in a perplexing way, an error will be thrown, which will be displayed in the results area.

bruth commented:
In your reading is there any support for the non-primitive types (arrays, structs, geo types)?

murphyke commented:
@bruth There is support for at least some of that when using the "standard" SQL dialect, which is what the JS API uses by default. Note that `bq` uses the "legacy" SQL dialect by default, and if you want to run standard queries with it, you have to do `bq ... query --nouse_legacy_sql ...`.

murphyke commented:
Yes, BQ standard SQL supports array, struct, and a [GEOGRAPHY](https://cloud.google.com/bigquery/docs/gis-data) data type as an extension.

bruth commented:
> Yes, BQ standard SQL supports..

Does SQLPad support rendering these types?

murphyke commented:
> > Yes, BQ standard SQL supports..
> 
> Does SQLPad support rendering these types?

It doesn't blow up.

In the tabular layout, arrays look like what you would expect:

```
SELECT struct(1 as a, 2 as b) as aStruct, [1,2,3] as anArray;

aStruct                 anArray
{ "a": 1, "b": 2 }    [ 1, 2, 3 ]
```

and in JSON:

`[{ "aStruct": { "a": 1, "b": 2 }, "anArray": [1, 2, 3] }]`

I saw an issue saying there is no support for them in the charts, but that doesn't seem so surprising.

murphyke commented:
BTW, I saw an issue from September about how the SQLPad author was abandoning the project, but somebody is now paying him to work on it for a couple weeks, followed by part-time for a couple months, and he has been a ball of fire.

murphyke commented:
AFAIK, I'm caught up on the problems now.

bruth commented:
Great that it supports those other data types.

> BTW, I saw an issue from September about how the SQLPad author was abandoning the project, but somebody is now paying him to work on it for a couple weeks, followed by part-time for a couple months, and he has been a ball of fire.

Interesting. Well the project seems in a good state so it feels fairly easy to maintain if need be, but it is good to know that he is open to paid support.

bruth commented:
> AFAIK, I'm caught up on the problems now.

Ok, I will approve and we can move forward since it is functional. Great job and timeliness of this.. this is much needed for the Labs.
